### DISCUSSION TITLE
What is the best way of having a job retry four times a day until a date?

### DISCUSSION THREAD

**Discussion by stoivo** [sidekiq/sidekiq#5397] [General]
Hi,

TLDR;  
What is the best way of having a job retry four times a day until a date?

I am looking for some suggestions or recommendations. I want to create a job that pulls the status of the payment after each time the bank is processing payments. `BankClearingCycle::Norway.new.next_clearing_cycle` does the work of finding the next next time.

The execution date can be today or it can be a year in the future (imagine someone registering all the interests for the entire year). 

I want it to pull the status until the payment stops being active or if we are 14 days after the execution date.

I started with pulling for 14 days. It looked like this
```ruby
class ZtlPollPaymentStatusJob
  include Sidekiq::Job

  # We want to try for the next 2 weeks after each clearing cycle
  sidekiq_options retry: BankClearingCycle::Norway.new.todays_times.size * 14
  sidekiq_retry_in do |_count, exception|
    next unless exception.is_a? CheckStatusAgainOnNextCycle

    Rails.logger.info {
      "will retry at next cycle #{BankClearingCycle::Norway.new.next_clearing_cycle} + 5 min"
    }
    # sidekiq_retry_in should return the int for seconds to wait before executing
    # again. First find seconds between next cycle and current time, add a 5
    # minutes delay since it takes some time from the cycle to to finish process
    # and some jitter
    seconds_to_next_cycle = BankClearingCycle::Norway.new.next_clearing_cycle - Time.zone.now
    (seconds_to_next_cycle + 5.minutes + rand(5.minutes)).to_i
  end

  class CheckStatusAgainOnNextCycle < StandardError
  end

  def perform(id)
    basket = ZtlPaymentBasket.find(id)
    Ztl::CheckPaymentStatus.call(basket)
    basket.reload

    Rails.logger.info { "status is #{basket.current_status.process_status}" }
    if basket.current_status.process_status == "active"
      raise CheckStatusAgainOnNextCycle
    end
  end
end
```

I think I can change it for:
```ruby
class ZtlPollPaymentStatusJob
  include Sidekiq::Job

  # We want to try for the next year after each clearing cycle
  sidekiq_options retry: BankClearingCycle::Norway.new.todays_times.size * 365
  sidekiq_retry_in do |_count, exception|
    next unless exception.is_a? CheckStatusAgainOnNextCycle

    Rails.logger.info {
      "will retry at next cycle #{BankClearingCycle::Norway.new.next_clearing_cycle} + 5 min"
    }
    # sidekiq_retry_in should return the int for seconds to wait before executing
    # again. First find seconds between next cycle and current time, add a 5
    # minutes delay since it takes some time from the cycle to to finish process
    # and some jitter
    seconds_to_next_cycle = BankClearingCycle::Norway.new.next_clearing_cycle - Time.zone.now
    (seconds_to_next_cycle + 5.minutes + rand(5.minutes)).to_i
  end

  class CheckStatusAgainOnNextCycle < StandardError
  end

  def perform(id)
    basket = ZtlPaymentBasket.find(id)
    Ztl::CheckPaymentStatus.call(basket)
    basket.reload

    Rails.logger.info { "status is #{basket.current_status.process_status}" }
    
    return if basket.current_status.process_status == "active"
    if (basket.payments.maximum(:execution_date) + 2.weeks).past?
      Rails.logger.info { "payment had execution_date 14 days ago so we stop to retry" }
      return 
    end
    
    raise CheckStatusAgainOnNextCycle
  end
end
```

My question: Is this an elegant solution or do you have some other recommendation?

**Comment by mperham** [comment]
Use sidekiq_retry_in { 6 * 60 * 60 } and have the job return after early
after the deadline.

**Reply by stoivo** [reply]
Ok, thanks. So something very similar to what I did. Thanks, Mike. You are amazing, relying within an hour on a sunday(at least here in Norway)

It needs to be a bit more complicated because the clearing cycles aren't evenly spaced
```
BankClearingCycle::Norway.new.todays_times.map{_1.strftime("%H:%M")}                                                                              
=> ["05:30", "09:30", "11:30", "13:30", "15:30"]
```

**Reply by stoivo** [reply]
Thanks. So something very similar to that I did. I have to say you are awesome, reply witin an hour on a sunday (at least here in Norway)

I have to do something more then `sidekiq_retry_in { 6 * 60 * 60 }` because the banking cycle is't evenly spread
```ruby
BankClearingCycle::Norway.new.todays_times.map{_1.strftime("%H:%M")}                                                                              
# => ["05:30", "09:30", "11:30", "13:30", "15:30"]
```

**Reply by stoivo** [reply]
Is there a way I can avoid logging the backtrace when I raise CheckStatusAgainOnNextCycle? I see that there are JobRetry::Handled and JobRetry::Skip, but those seams to do other things. 

This works but feels a but dirty
```rb
  class CheckStatusAgainOnNextCycle < StandardError
    def skip_bugsnag
      true
    end

    def backtrace; nil end
  end
```

**Comment by mperham** [comment]
Nope. If the job raises an error, Sidekiq will log the backtrace. There are
workarounds as you’ve discovered. You can also replace the default error
handler.

**Reply by stoivo** [reply]
Ahh, clever. Like remove Sidekiq::ExceptionHandler::Logger from the error_handlers and add one of my own.

**Comment by mperham** [comment]
Right. I could make improvements for 7.0 but this is a slippery slope. As
soon as you say “don’t log this error” you’ll come up with another edge
case where you do want to log the backtrace. That’s why I have the current
policy on the error handling wiki page.

**Comment by stoivo** [comment]
Are there any other ways to retry a job without raiseing an Exception?   
I see that `JobRetry::Handled` and `JobRetry::Skip` exists but what are those for?

**Comment by mperham** [comment]
They are internal details due to the two-stage local and global handlers
within job_retry.rb. You can use Skip but it’s not an official public API.

**Reply by stoivo** [reply]
I will take that as: you have to raise exception to rety.

**Comment by stoivo** [comment]
Sorry for going a bit back and fourth.

Do you have any opinion on if it's better to retry a job or schedule a new job
with the same arguments?

I talked with some of my colleagues that suggested scheduling a new job instead
of retrying the same job. Then we would not need to overwrite sidekiq_retry_in
or error_handler.


### retry the same job
```rb
class ZtlPollPaymentStatusJob
  include Sidekiq::Job

  # We want to try for the next year after each clearing cycle
  sidekiq_options retry: BankClearingCycle::Norway.new.todays_times.size * 365
  sidekiq_retry_in do |_count, exception|
    next unless exception.is_a? CheckStatusAgainOnNextCycle

    Rails.logger.info {
      "will retry at next cycle #{BankClearingCycle::Norway.new.next_clearing_cycle} + 5 min"
    }
    # sidekiq_retry_in should return the int for seconds to wait before executing
    # again. First find seconds between next cycle and current time, add a 5
    # minutes delay since it takes some time from the cycle to to finish process
    # and some jitter
    seconds_to_next_cycle = BankClearingCycle::Norway.new.next_clearing_cycle - Time.zone.now
    (seconds_to_next_cycle + 5.minutes + rand(5.minutes)).to_i
  end

  class CheckStatusAgainOnNextCycle < StandardError
  end

  def perform(id)
    basket = ZtlPaymentBasket.find(id)
    Ztl::CheckPaymentStatus.call(basket)
    basket.reload

    Rails.logger.info { "status is #{basket.current_status.process_status}" }
    
    return if basket.current_status.process_status == "active"
    if (basket.payments.maximum(:execution_date) + 2.weeks).past?
      Rails.logger.info { "payment had execution_date 14 days ago so we stop to retry" }
      return 
    end
    
    raise CheckStatusAgainOnNextCycle
  end
end
```

### schedule a new job
```rb
class ZtlPollPaymentStatusJob
  include Sidekiq::Job

  def perform(id)
    basket = ZtlPaymentBasket.find(id)
    Ztl::CheckPaymentStatus.call(basket)
    basket.reload

    Rails.logger.info { "status is #{basket.current_status.process_status}" }
    
    return if basket.current_status.process_status == "active"
    if (basket.payments.maximum(:execution_date) + 2.weeks).past?
      Rails.logger.info { "payment had execution_date 14 days ago so we stop to retry" }
      return 
    end
    
    ZtlPollPaymentStatusJob
      .perform_at(BankClearingCycle::Norway.new.next_clearing_cycle + 5.minutes + rand(5.minutes), id)
  end
end
```

**Comment by mperham** [comment]
I think rescheduling the job is possibly a better solution. Exceptions
should be unexpected issues. If something is expected, you should handle it
as normal. If the job runs out of “reschedules”, you could then start
raising errors.
